Sign fetched block

hewison-chris opened the pull request:
If a block is fetched from another node it is better to make sure our signature is in the block.

hewison-chris committed:
Trivial: improvement to logging of sig catchup

hewison-chris committed:
Trivial: remove misleading log line

It implied that the block had been added to the ledger which is not the
case unless the majority have already signed.

hewison-chris committed:
Validator: self sign blocks received

If as a validator we fetch a block from another node during catchup it
may not be self signed yet. We should add our signature in this case.

mkykadir commented on source/agora/node/Validator.d:
```diff
@@ -438,11 +438,19 @@ public class Validator : FullNode, API
 
     protected override string acceptHeader (BlockHeader header) @safe
     {
-        if (auto err = super.acceptHeader(header))
```
`FullNode.acceptHeader` also invoked Ledger signature update, but we're missing it right now.

hewison-chris replied:
It will still update the signatures when `selfSignBlock` is called.

mkykadir replied:
I might be missing something but; couldn't reach to `this.storage.updateBlockSig` that we had from new changeset

hewison-chris replied:
Sorry I see what you mean now. I will add the `this.ledger.updateBlockMultiSig(header);` in `Validator`.

hewison-chris replied:
Pushed an update.

hewison-chris committed:
Trivial replacement of already calculated value

hewison-chris committed:
Trivial debug logging with function name

hewison-chris committed:
PeriodicCatchup unit test updated to override checkExternalize

Since the `valueExternalized` is no longer when the block is actually
externalized to the ledger, the test now overrides `checkExternalize`.

hewison-chris committed:
Trivial info log addition

hewison-chris committed:
Trivial debug logging with function name

hewison-chris committed:
Resend latest envelopes if half block late

Rather than waiting a full block interval to start resending envelopes
with same nomination it can help prevent delayed blocks if we only wait
for half the duration. This will probably need to be optimized at some
point but first we want to ensure we do not get blocked when peers are
not connected when we first sent the envelopes.

hewison-chris committed:
System Integration test: Increase frequency of block catchup

Now that blocks are getting externalized faster the nodes should check
for catchup more frequently. This is especially important for the
fullnode which has been seen to get many blocks behind otherwise.

hewison-chris committed:
Unit tests: Check more frequently for blocks to catchup

hewison-chris commented:
All green:-)

hewison-chris commented on source/agora/node/Validator.d:
```diff
@@ -441,11 +441,20 @@ public class Validator : FullNode, API
+        if (header.height > this.ledger.getLastPaidHeight())
+        {
+            this.nominator.selfSignBlock(header);
+            this.ledger.updateBlockMultiSig(header);
```
@mkykadir Added it here.

linked0 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -600,14 +600,16 @@ extern(D):
-            if (expected > this.ledger.height + 1) // We are one block late
+            const Duration halfBlockDuration = this.params.BlockInterval / 2;
+            // We are half way through block interval time to externalize
+            if (this.clock.utcTime() > this.ledger.getExpectedBlockTime(this.ledger.height) + halfBlockDuration)
```
```suggestion
            if (this.clock.utcTime() > this.ledger.getExpectedBlockTime(this.ledger.height) + 
                this.params.BlockInterval / 2)
```
Can't we just use this operation without introducing the variable?

hewison-chris replied:
The `const` is just to self document.

linked0 reviewed (approved):
LGTM. My comment doesn't need to CI test of course. :)
